src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,465 @@
+
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error_errno(r, "Can't get bpf prog info for fd %i: %m", fd);
+                        continue;
+                }
+
+                /* If there are maps associated with this prog, fetch bpf_prog_info again
+                 * to get a list of map ids */
+                nr_map_ids = prog_info.nr_map_ids;
+                if (nr_map_ids > 0) {
+                        prog_map_ids = new(uint32_t, nr_map_ids);
+                        if (!prog_map_ids)
+                                return -ENOMEM;
+                        prog_map_names = new(char, nr_map_ids * BPF_OBJ_NAME_LEN);
+                        if (!prog_map_names)
+                                return -ENOMEM;
+
+                        zero(prog_info);
+                        prog_info = (struct bpf_prog_info) {
+                                .nr_map_ids = nr_map_ids,
+                                .map_ids = PTR_TO_UINT64(prog_map_ids),
+                        };
+                        r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                        if (r < 0) {
+                                log_error_errno(r, "Can't get bpf prog info for fd %i: %m", fd);
+                                continue;
+                        }
+
+                        /* For each map id, get the map name from the hashmap and concatenate them
+                         * into one string, separated by colons. */
+                        i = 0;
+                        FOREACH_ARRAY(j, prog_map_ids, nr_map_ids) {
+                                uint32_t map_id = *j;
+                                char *map_name = hashmap_get(map_name_by_id, UINT32_TO_PTR(map_id));
+                                if (map_name == NULL) {
+                                        log_error_errno(r, "Can't find map in map_name_by_id %u: %m", map_id);
+                                        continue;
+                                }
+                                strcpy(&prog_map_names[i], map_name);
+                                i += strlen(map_name);
+                                prog_map_names[i] = ':';
+                                i++;
+                        }
+                        prog_map_names[i] = '\0';
+                }
+
+                /* Fetch memlock using fd from /proc/self/fdinfo */
+                r = get_memlock_by_fd(fd, &memlock);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read memlock from /proc/self/fdinfo/%i", fd);
+                        continue;
+                }
+
+                /* Format prog tag */
+                prog_tag[BPF_TAG_SIZE * 2] = '\0';
+                for (size_t j = 0; j < BPF_TAG_SIZE; ++j)
+                        sprintf(&(prog_tag[2 * j]), "%02hhx", prog_info.tag[j]);
```

Comment:
```c
prog_tag = hexmem(prog_info.tag, BPF_TAG_SIZE);
if (!prog_tag)
        return log_oom();
```

and this use dynamic memory.

in systemd we really care about keeping things easy to read. I doubt this is so performance sensitive that a malloc() here would actually hurt